edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs

```diff
@@ -0,0 +1,693 @@
+                if (!authorized)
+                {
+                    await this.SendResponseAsync(HttpStatusCode.Unauthorized);
+                    return;
+                }
+
+                Events.Authorized(nameof(this.CreateOrUpdateModuleOnBehalfOfAsync), actorDeviceId, Constants.EdgeHubModuleId, targetId, authChain);
+                string edgeDeviceId = edgeHub.GetEdgeDeviceId();
+                RegistryApiHttpResult result = await this.apiClient.PutModuleAsync(
+                    edgeDeviceId,
+                    new CreateOrUpdateModuleOnBehalfOfData($"{AuthChainHelpers.SkipFirstIdentityFromAuthChain(authChain)}", requestData.Module),
+                    ifMatchHeader);
+                await this.SendResponseAsync(result.StatusCode, result.JsonContent);
+                Events.CompleteRequest(nameof(this.CreateOrUpdateModuleOnBehalfOfAsync), edgeDeviceId, targetId, requestData.AuthChain, result);
+            }
+            catch (Exception ex)
+            {
+                Events.InternalServerError(nameof(this.CreateOrUpdateModuleOnBehalfOfAsync), ex);
+                await this.SendResponseAsync(HttpStatusCode.InternalServerError, FormatErrorResponseMessage(ex.ToString()));
+            }
+        }
+
+        [HttpGet]
+        [Route("devices/{deviceId}/modules/{moduleId}")]
+        public async Task GetModuleAsync(
+            [FromRoute] string deviceId,
+            [FromRoute] string moduleId)
+        {
+            try
+            {
+                Events.ReceivedRequest(nameof(this.GetModuleAsync), deviceId, moduleId);
+
+                try
+                {
+                    deviceId = WebUtility.UrlDecode(Preconditions.CheckNonWhiteSpace(deviceId, nameof(deviceId)));
+                    moduleId = WebUtility.UrlDecode(Preconditions.CheckNonWhiteSpace(moduleId, nameof(moduleId)));
+                }
+                catch (Exception ex)
+                {
+                    Events.BadRequest(nameof(this.GetModuleAsync), ex.Message);
+                    await this.SendResponseAsync(HttpStatusCode.BadRequest, FormatErrorResponseMessage(ex.Message));
+                    return;
+                }
+
+                if (!await this.AuthenticateAsync(deviceId, Option.None<string>()))
+                {
+                    await this.SendResponseAsync(HttpStatusCode.Unauthorized);
+                    return;
+                }
+
+                IEdgeHub edgeHub = await this.edgeHubGetter;
+                string edgeDeviceId = edgeHub.GetEdgeDeviceId();
+                string targetId = $"{deviceId}/{moduleId}";
+                IDeviceScopeIdentitiesCache identitiesCache = edgeHub.GetDeviceScopeIdentitiesCache();
+
+                (bool authorized, string authChain) = await this.AuthorizeActorAsync(identitiesCache, edgeDeviceId, Constants.EdgeHubModuleId, targetId);
```

Comment:
Who are we authorizing here? And why? 
In this scenario, the child edge device is making a call to getModule, and the current edge device will make the on-behalf-of call, right? Are we validating if the device is our child? If so, won't Authenticate do that?

Reply:
yes, this will validate target device is current edge's child and help to return authChain (which is required to use for on-behalf-of api call).  For authenticate we are verifying child device sas token or certificate is valid.

Reply:
Authenticate already handles that today for existing API calls.  But it's fine for for now as I need to come back and clean out all duplidated AuthZ implementations from both controllers (see other comment).

---
In reply to: [490649553](https://github.com/Azure/iotedge/pull/3532#discussion_r490649553) [](ancestors = 490649553)